File: rust-api-examples/examples/cohere_transcribe.rs
// Copyright (c) 2026 Xiaomi Corporation
//
// This file demonstrates how to use Cohere Transcribe with sherpa-onnx's Rust
// API for offline speech recognition.

use clap::Parser;
use sherpa_onnx::{
    OfflineCohereTranscribeModelConfig, OfflineRecognizer, OfflineRecognizerConfig, Wave,
};

/// Cohere Transcribe offline example
#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
    /// Path to WAV file
    #[arg(long)]
    wav: String,

    /// Path to encoder ONNX model
    #[arg(long)]
    encoder: String,

    /// Path to decoder ONNX model
    #[arg(long)]
    decoder: String,

    /// Path to tokens.txt
    #[arg(long)]
    tokens: String,

    /// Language to decode
    #[arg(long, default_value = "en")]
    language: String,

    /// Provider (default: cpu)
    #[arg(long, default_value = "cpu")]
    provider: String,

    /// Enable debug logs
    #[arg(long, default_value_t = false)]
    debug: bool,

    /// Number of threads
    #[arg(long, default_value_t = 2)]
    num_threads: i32,
}

fn main() {
    let args = Args::parse();

    let wave = Wave::read(&args.wav).expect("Failed to read WAV file");

    let mut recognizer_config = OfflineRecognizerConfig::default();

    recognizer_config.model_config.cohere_transcribe = OfflineCohereTranscribeModelConfig {
        encoder: Some(args.encoder.clone()),
        decoder: Some(args.decoder.clone()),
        use_punct: true,
        use_itn: true,
        ..Default::default()
    };

    recognizer_config.model_config.tokens = Some(args.tokens.clone());
    recognizer_config.model_config.provider = Some(args.provider.clone());
    recognizer_config.model_config.debug = args.debug;
    recognizer_config.model_config.num_threads = args.num_threads;

    let recognizer =
        OfflineRecognizer::create(&recognizer_config).expect("Failed to create OfflineRecognizer");
    let stream = recognizer.create_stream();
    stream.set_option("language", &args.language);
    stream.accept_waveform(wave.sample_rate(), wave.samples());
    recognizer.decode(&stream);

    if let Some(result) = stream.get_result() {
        println!("Decoded text: {}", result.text);
    } else {
        eprintln!("Failed to get recognition result");
    }
}

File: dart-api-examples/non-streaming-asr/bin/cohere-transcribe.dart
// Copyright (c)  2026  Xiaomi Corporation
import 'dart:io';

import 'package:args/args.dart';
import 'package:sherpa_onnx/sherpa_onnx.dart' as sherpa_onnx;

import './init.dart';

void main(List<String> arguments) async {
  await initSherpaOnnx();

  final parser = ArgParser()
    ..addOption('encoder', help: 'Path to the encoder model')
    ..addOption('decoder', help: 'Path to the decoder model')
    ..addOption('tokens', help: 'Path to tokens.txt')
    ..addOption('input-wav', help: 'Path to input.wav to transcribe')
    ..addOption('language', help: 'Language to decode', defaultsTo: 'en');

  final res = parser.parse(arguments);
  if (res['encoder'] == null ||
      res['decoder'] == null ||
      res['tokens'] == null ||
      res['input-wav'] == null) {
    print(parser.usage);
    exit(1);
  }

  final encoder = res['encoder'] as String;
  final decoder = res['decoder'] as String;
  final tokens = res['tokens'] as String;
  final inputWav = res['input-wav'] as String;
  final language = res['language'] as String;

  final cohereTranscribe = sherpa_onnx.OfflineCohereTranscribeModelConfig(
    encoder: encoder,
    decoder: decoder,
    usePunct: true,
    useItn: true,
  );

  final modelConfig = sherpa_onnx.OfflineModelConfig(
    cohereTranscribe: cohereTranscribe,
    tokens: tokens,
    debug: true,
    numThreads: 1,
  );
  final config = sherpa_onnx.OfflineRecognizerConfig(model: modelConfig);
  final recognizer = sherpa_onnx.OfflineRecognizer(config);

  final waveData = sherpa_onnx.readWave(inputWav);
  final stream = recognizer.createStream();
  stream.setOption(key: 'language', value: language);
  stream.acceptWaveform(
    samples: waveData.samples,
    sampleRate: waveData.sampleRate,
  );

  recognizer.decode(stream);

  final result = recognizer.getResult(stream);
  print(result.text);

  stream.free();
  recognizer.free();
}
